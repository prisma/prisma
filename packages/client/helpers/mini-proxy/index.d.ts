/// <reference types="node" />

import https from 'https';

export declare type CertificatesConfig = {
    caKey: string;
    caCert: string;
    key: string;
    cert: string;
    csr: string;
};

export declare type ConnectionStringConfig = {
    databaseUrl: string;
    envVar: string;
    port: number;
};

export declare function decodeKey(key: string): KeyContents;

export declare function decodeKeyOrNull(key: string): KeyContents | null;

export declare const defaultCertificatesConfig: CertificatesConfig;

export declare const defaultServerConfig: ServerConfig;

export declare function encodeKey(config: EncodeKeyConfig): string;

export declare type EncodeKeyConfig = {
    databaseUrl: string;
    envVar: string;
};

export declare function generateCertificates(config: CertificatesConfig): Promise<void>;

export declare function generateConnectionString(config: ConnectionStringConfig): string;

export declare type KeyContents = {
    url: string;
    envVar: string;
};

export declare type ServerConfig = {
    platform?: string;
    queryEngine?: string;
    port: number;
    ca?: string;
    key: string;
    cert: string;
};

export declare function startServer(config: ServerConfig): Promise<https.Server>;

export { }
