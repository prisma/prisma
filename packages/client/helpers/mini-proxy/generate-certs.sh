#!/bin/bash

set -euxo pipefail

if [[ $# -ne 5 ]]; then
  >&2 echo "usage: generate-certs.sh CA_KEY CA_CERT CSR KEY CERT"
  exit 1
fi

CA_KEY=$1
CA_CERT=$2
CSR=$3
KEY=$4
CERT=$5

openssl req -x509 -newkey rsa:4096 -nodes -sha256 -days 3650 \
  -keyout "$CA_KEY" -out "$CA_CERT" \
  -subj "/O=Prisma"

openssl req -newkey rsa:4096 -nodes \
  -keyout "$KEY" -out "$CSR" \
  -subj "/CN=localhost" \
  -extensions san -config <(echo '[req]';
                            echo 'distinguished_name=req';
                            echo '[san]';
                            echo 'subjectAltName=DNS:localhost,IP:127.0.0.1')

openssl x509 -req -sha256 -days 3650 \
  -in "$CSR" -out "$CERT" \
  -CA "$CA_CERT" -CAkey "$CA_KEY" -CAcreateserial
