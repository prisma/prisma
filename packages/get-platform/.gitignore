node_modules
.DS_Store
query-engine*
yarn-error.log
pnpm-debug.log

*.db

dist/

copy-photon.sh
*.tsbuildinfo