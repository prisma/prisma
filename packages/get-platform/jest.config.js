module.exports = {
  transform: {
    '^.+\\.(m?j|t)s$': '@swc/jest',
  },
  transformIgnorePatterns: [],
  testEnvironment: 'node',
  collectCoverage: process.env.CI ? true : false,
  coverageReporters: ['clover'],
  coverageDirectory: 'src/__tests__/coverage',
  collectCoverageFrom: ['src/**/*.ts', '!**/__tests__/**/*'],
  testMatch: ['**/src/__tests__/**/*.test.ts'],
}
