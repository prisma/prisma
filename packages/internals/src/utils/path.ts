import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
