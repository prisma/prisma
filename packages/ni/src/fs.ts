import { resolve } from 'node:path'
import fs from 'node:fs'
import type { RunnerContext } from './runner'

export function getPackageJSON(ctx?: RunnerContext): any {
  const cwd = ctx?.cwd ?? process.cwd()
  const path = resolve(cwd, 'package.json')

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw)
      return data
    }
    catch (e) {
      if (!ctx?.programmatic) {
        console.warn('Failed to parse package.json')
        process.exit(1)
      }

      throw e
    }
  }
}
