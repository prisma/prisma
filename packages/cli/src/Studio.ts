import Debug from '@prisma/debug'
import { enginesVersion } from '@prisma/engines'
import {
  arg,
  Command,
  format,
  getConfig,
  getDirectUrl,
  HelpError,
  isError,
  loadEnvFile,
  resolveUrl,
} from '@prisma/internals'
import { getSchemaPathAndPrint } from '@prisma/migrate'
import { StudioServer } from '@prisma/studio-server'
import fs from 'fs'
import getPort from 'get-port'
import { bold, dim, red } from 'kleur/colors'
import open from 'open'
import path from 'path'

// Note that we have a test relying on the namespace
// Any change to the namespace must be done in the test as well
// See packages/client/tests/e2e/issues/studio-1128-spawn-enoent/_steps.ts
const debug = Debug('prisma:cli:studio')

const packageJson = require('../package.json') // eslint-disable-line @typescript-eslint/no-var-requires

export class Studio implements Command {
  public instance?: StudioServer

  public static new(): Studio {
    return new Studio()
  }

  private static help = format(`
Browse your data with Prisma Studio

${bold('Usage')}

  ${dim('$')} prisma studio [options]

${bold('Options')}

  -h, --help        Display this help message
  -p, --port        Port to start Studio on
  -b, --browser     Browser to open Studio in
  -n, --hostname    Hostname to bind the Express server to
  --schema          Custom path to your Prisma schema

${bold('Examples')}

  Start Studio on the default port
    ${dim('$')} prisma studio

  Start Studio on a custom port
    ${dim('$')} prisma studio --port 5555

  Start Studio in a specific browser
    ${dim('$')} prisma studio --port 5555 --browser firefox
    ${dim('$')} BROWSER=firefox prisma studio --port 5555

  Start Studio without opening in a browser
    ${dim('$')} prisma studio --port 5555 --browser none
    ${dim('$')} BROWSER=none prisma studio --port 5555

  Specify a schema
    ${dim('$')} prisma studio --schema=./schema.prisma
`)

  /**
   * Parses arguments passed to this command, and starts Studio
   *
   * @param argv Array of all arguments
   */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--port': Number,
      '-p': '--port',
      '--browser': String,
      '-b': '--browser',
      '--hostname': String,
      '-n': '--hostname',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    loadEnvFile({ schemaPath: args['--schema'], printMessage: true })

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const hostname = args['--hostname']
    const port = args['--port'] || (await getPort({ port: getPort.makeRange(5555, 5600) }))
    const browser = args['--browser'] || process.env.BROWSER

    const staticAssetDir = path.resolve(__dirname, '../build/public')

    const schema = await fs.promises.readFile(schemaPath, 'utf-8')
    const config = await getConfig({ datamodel: schema, ignoreEnvVarErrors: true })

    process.env.PRISMA_DISABLE_WARNINGS = 'true' // disable client warnings
    const studio = new StudioServer({
      schemaPath,
      hostname,
      port,
      staticAssetDir,
      prismaClient: {
        resolve: {
          '@prisma/client': path.resolve(__dirname, '../prisma-client/index.js'),
        },
        directUrl: resolveUrl(getDirectUrl(config.datasources[0])),
      },
      versions: {
        prisma: packageJson.version,
        queryEngine: enginesVersion,
      },
    })

    await studio.start()

    const serverUrl = `http://localhost:${port}`
    if (!browser || browser.toLowerCase() !== 'none') {
      try {
        const subprocess = await open(serverUrl, {
          app: browser,
          url: true,
        })

        subprocess.on('spawn', () => {
          // We match on this string in debug logs in tests
          debug(`requested to open the url ${serverUrl}`)
        })

        subprocess.on('error', (e) => {
          debug(e)
          // We match on this string in debug logs in tests
          debug(`failed to open the url ${serverUrl} in browser`)
        })
      } catch (e) {
        // Ignore any errors that occur when trying to open the browser, since they should not halt the process
        debug(e)
      }
    }

    this.instance = studio

    return `Prisma Studio is up on ${serverUrl}`
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${bold(red(`!`))} ${error}\n${Studio.help}`)
    }

    return Studio.help
  }
}
