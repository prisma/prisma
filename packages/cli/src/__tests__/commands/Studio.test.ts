import { jestConsoleContext, jestContext } from '@prisma/get-platform'
import * as miniProxy from '@prisma/mini-proxy'
import fs from 'fs'
import fetch from 'node-fetch'
import path from 'path'
import rimraf from 'rimraf'

import { DbPush } from '../../../../migrate/src/commands/DbPush'
import { Studio } from '../../Studio'

const originalEnv = { ...process.env }

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

const STUDIO_TEST_PORT = 5678

const testIf = (condition: boolean) => (condition ? test : test.skip)
const describeIf = (condition: boolean) => (condition ? describe : describe.skip)

function sendRequest(message: any): Promise<any> {
  return fetch(`http://localhost:${STUDIO_TEST_PORT}/api`, {
    method: 'POST',
    headers: {
      'Content-Type': 'text/plain',
    },
    body: JSON.stringify(message),
  }).then((res) => res.json())
}

let studio: Studio
// Prisma Studio ignores env vars for overriding engine paths, skipping test for now
describeIf(!process.env.PRISMA_QUERY_ENGINE_LIBRARY && !process.env.PRISMA_QUERY_ENGINE_BINARY)(
  'studio with alternative urls and prisma://',
  () => {
    afterEach(() => {
      // Back to original env vars
      process.env = { ...originalEnv }
    })

    test('queries work if url is prisma:// and directUrl is set', async () => {
      process.env.PDP_URL = 'prisma://aws-us-east-1.prisma-data.com/?api_key=MY_API_KEY'
      process.env.DATABASE_URL = process.env.TEST_POSTGRES_URI!.replace('tests', `tests-${Date.now()}-studio`)

      ctx.fixture('schema-only-data-proxy-direct-url')

      const studio = Studio.new()

      await DbPush.new().parse(['--schema', 'schema.prisma', '--skip-generate'])
      const result = studio.parse(['--port', `${STUDIO_TEST_PORT}`, '--browser', 'none'])

      await expect(result).resolves.not.toThrow()

      const res = await sendRequest({
        requestId: 1,
        channel: 'prisma',
        action: 'clientRequest',
        payload: {
          data: {
            modelName: 'SomeUser',
            operation: 'findMany',
            args: {
              select: {
                id: true,
              },
            },
          },
        },
      })

      expect(res).toMatchSnapshot()

      studio.instance?.stop()
    })

    testIf(process.platform !== 'win32')('queries work if url is prisma:// via the mini-proxy', async () => {
      process.env.DATABASE_URL = process.env.TEST_POSTGRES_URI!.replace('tests', `tests-${Date.now()}-studio`)
      process.env.PDP_URL = miniProxy.generateConnectionString({
        envVar: 'PDP_URL',
        databaseUrl: process.env.DATABASE_URL,
        port: miniProxy.defaultServerConfig.port,
      })

      ctx.fixture('schema-only-data-proxy')

      await DbPush.new().parse(['--schema', 'schema.prisma', '--skip-generate'])
      delete process.env.DATABASE_URL

      const studio = Studio.new()
      const result = studio.parse(['--port', `${STUDIO_TEST_PORT}`, '--browser', 'none'])

      await expect(result).resolves.not.toThrow()

      const res = await sendRequest({
        requestId: 1,
        channel: 'prisma',
        action: 'clientRequest',
        payload: {
          data: {
            modelName: 'SomeUser',
            operation: 'findMany',
            args: {
              select: {
                id: true,
              },
            },
          },
        },
      })

      expect(res).toMatchSnapshot()

      studio.instance?.stop()

      expect(ctx.mocked['console.warn'].mock.calls.join('\n')).toMatchInlineSnapshot(`""`)
    })
  },
)

describe('studio with default schema.prisma filename', () => {
  jest.setTimeout(20_000)

  beforeAll(async () => {
    // Before every test, we'd like to reset the DB.
    // We do this by duplicating the original SQLite DB file, and using the duplicate as the datasource in our schema
    rimraf.sync(path.join(__dirname, '../fixtures/studio-test-project/dev_tmp.db'))
    fs.copyFileSync(
      path.join(__dirname, '../fixtures/studio-test-project/dev.db'),
      path.join(__dirname, '../fixtures/studio-test-project/dev_tmp.db'),
    )

    // Clean up Client generation directory
    rimraf.sync(path.join(__dirname, '../prisma-client'))
    studio = Studio.new()

    await studio.parse([
      '--schema',
      path.join(__dirname, '../fixtures/studio-test-project/schema.prisma'),
      '--port',
      `${STUDIO_TEST_PORT}`,
      '--browser',
      'none',
    ])

    // Give Studio time to start
    await new Promise((r) => setTimeout(() => r(null), 2_000))
  })

  afterAll(() => {
    studio.instance!.stop()
  })

  test('can start up correctly', async () => {
    const res = await fetch(`http://localhost:${STUDIO_TEST_PORT}`)
    expect(res.status).toEqual(200)
  })

  test('can respond to `findMany` queries', async () => {
    const res = await sendRequest({
      requestId: 1,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'findMany',
          args: {
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `create` queries', async () => {
    const res = await sendRequest({
      requestId: 2,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'create',
          args: {
            data: {
              id: 3,
              string: '',
              int: 0,
              float: 0.0,
              datetime: '2020-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `update` queries', async () => {
    const res = await sendRequest({
      requestId: 3,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'update',
          args: {
            where: {
              id: 1,
            },
            data: {
              string: 'Changed String',
              int: 100,
              float: 100.5,
              datetime: '2025-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `delete` queries', async () => {
    const res = await sendRequest({
      requestId: 4,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'delete',
          args: {
            where: { id: 2 },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })
    expect(res).toMatchSnapshot()
  })
})

describe('studio with custom schema.prisma filename', () => {
  jest.setTimeout(20_000)

  beforeAll(async () => {
    // Before every test, we'd like to reset the DB.
    // We do this by duplicating the original SQLite DB file, and using the duplicate as the datasource in our schema
    rimraf.sync(path.join(__dirname, '../fixtures/studio-test-project-custom-filename/dev_tmp.db'))
    fs.copyFileSync(
      path.join(__dirname, '../fixtures/studio-test-project-custom-filename/dev.db'),
      path.join(__dirname, '../fixtures/studio-test-project-custom-filename/dev_tmp.db'),
    )

    // Clean up Client generation directory
    rimraf.sync(path.join(__dirname, '../prisma-client'))
    studio = Studio.new()

    await studio.parse([
      '--schema',
      path.join(__dirname, '../fixtures/studio-test-project-custom-filename/schema1.prisma'),
      '--port',
      `${STUDIO_TEST_PORT}`,
      '--browser',
      'none',
    ])

    // Give Studio time to start
    await new Promise((r) => setTimeout(() => r(null), 2_000))
  })

  afterAll(() => {
    studio.instance!.stop()
  })

  test('can start up correctly', async () => {
    const res = await fetch(`http://localhost:${STUDIO_TEST_PORT}`)
    expect(res.status).toEqual(200)
  })

  test('can respond to `findMany` queries', async () => {
    const res = await sendRequest({
      requestId: 1,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'findMany',
          args: {
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `create` queries', async () => {
    const res = await sendRequest({
      requestId: 2,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'create',
          args: {
            data: {
              id: 3,
              string: '',
              int: 0,
              float: 0.0,
              datetime: '2020-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `update` queries', async () => {
    const res = await sendRequest({
      requestId: 3,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'update',
          args: {
            where: {
              id: 1,
            },
            data: {
              string: 'Changed String',
              int: 100,
              float: 100.5,
              datetime: '2025-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `delete` queries', async () => {
    const res = await sendRequest({
      requestId: 4,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'delete',
          args: {
            where: { id: 2 },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })
    expect(res).toMatchSnapshot()
  })
})

describe('studio with schema folder', () => {
  jest.setTimeout(20_000)

  beforeAll(async () => {
    // Before every test, we'd like to reset the DB.
    // We do this by duplicating the original SQLite DB file, and using the duplicate as the datasource in our schema
    rimraf.sync(path.join(__dirname, '../fixtures/studio-test-project-schema-folder/dev_tmp.db'))
    fs.copyFileSync(
      path.join(__dirname, '../fixtures/studio-test-project-schema-folder/dev.db'),
      path.join(__dirname, '../fixtures/studio-test-project-schema-folder/dev_tmp.db'),
    )

    // Clean up Client generation directory
    rimraf.sync(path.join(__dirname, '../prisma-client'))
    studio = Studio.new()

    await studio.parse([
      '--schema',
      path.join(__dirname, '../fixtures/studio-test-project-schema-folder/schema'),
      '--port',
      `${STUDIO_TEST_PORT}`,
      '--browser',
      'none',
    ])

    // Give Studio time to start
    await new Promise((r) => setTimeout(() => r(null), 2_000))
  })

  afterAll(() => {
    studio.instance!.stop()
  })

  test('can start up correctly', async () => {
    const res = await fetch(`http://localhost:${STUDIO_TEST_PORT}`)
    expect(res.status).toEqual(200)
  })

  test('can respond to `findMany` queries', async () => {
    const res = await sendRequest({
      requestId: 1,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'findMany',
          args: {
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `create` queries', async () => {
    const res = await sendRequest({
      requestId: 2,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'create',
          args: {
            data: {
              id: 3,
              string: '',
              int: 0,
              float: 0.0,
              datetime: '2020-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `update` queries', async () => {
    const res = await sendRequest({
      requestId: 3,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'update',
          args: {
            where: {
              id: 1,
            },
            data: {
              string: 'Changed String',
              int: 100,
              float: 100.5,
              datetime: '2025-08-03T00:00:00.000Z',
              relation: {
                connect: {
                  id: 3,
                },
              },
              relation_list: {
                connect: {
                  id: 3,
                },
              },
            },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  test('can respond to `delete` queries', async () => {
    const res = await sendRequest({
      requestId: 4,
      channel: 'prisma',
      action: 'clientRequest',
      payload: {
        data: {
          modelName: 'with_all_field_types',
          operation: 'delete',
          args: {
            where: { id: 2 },
            select: {
              id: true,
              string: true,
              int: true,
              float: true,
              datetime: true,
              relation: true,
              relation_list: true,
            },
          },
        },
      },
    })
    expect(res).toMatchSnapshot()
  })
})
