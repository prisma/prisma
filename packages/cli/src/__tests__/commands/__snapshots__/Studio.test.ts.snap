// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`studio with custom schema.prisma filename can respond to \`create\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 2020-08-03T00:00:00.000Z,
      float: 0,
      id: 3,
      int: 0,
      relation: {
        id: 3,
        name: Relation Target 003,
        waft_id: 3,
      },
      relation_list: [
        {
          id: 3,
          name: Relation Target 003,
          waft_id: 3,
        },
      ],
      string: ,
    },
    error: null,
  },
  requestId: 2,
}
`;

exports[`studio with custom schema.prisma filename can respond to \`delete\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 1970-01-01T00:00:00.000Z,
      float: 0,
      id: 2,
      int: 0,
      relation: {
        id: 2,
        name: Relation Target 002,
        waft_id: 2,
      },
      relation_list: [],
      string: Delete me,
    },
    error: null,
  },
  requestId: 4,
}
`;

exports[`studio with custom schema.prisma filename can respond to \`findMany\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: [
      {
        datetime: 2020-08-03T00:00:00.000Z,
        float: 3.14,
        id: 1,
        int: 42,
        relation: {
          id: 1,
          name: Relation Target 001,
          waft_id: 1,
        },
        relation_list: [],
        string: Some string,
      },
      {
        datetime: 1970-01-01T00:00:00.000Z,
        float: 0,
        id: 2,
        int: 0,
        relation: {
          id: 2,
          name: Relation Target 002,
          waft_id: 2,
        },
        relation_list: [],
        string: Delete me,
      },
    ],
    error: null,
  },
  requestId: 1,
}
`;

exports[`studio with custom schema.prisma filename can respond to \`update\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 2025-08-03T00:00:00.000Z,
      float: 100.5,
      id: 1,
      int: 100,
      relation: {
        id: 3,
        name: Relation Target 003,
        waft_id: 1,
      },
      relation_list: [
        {
          id: 3,
          name: Relation Target 003,
          waft_id: 1,
        },
      ],
      string: Changed String,
    },
    error: null,
  },
  requestId: 3,
}
`;

exports[`studio with default schema.prisma filename can respond to \`create\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 2020-08-03T00:00:00.000Z,
      float: 0,
      id: 3,
      int: 0,
      relation: {
        id: 3,
        name: Relation Target 003,
        waft_id: 3,
      },
      relation_list: [
        {
          id: 3,
          name: Relation Target 003,
          waft_id: 3,
        },
      ],
      string: ,
    },
    error: null,
  },
  requestId: 2,
}
`;

exports[`studio with default schema.prisma filename can respond to \`delete\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 1970-01-01T00:00:00.000Z,
      float: 0,
      id: 2,
      int: 0,
      relation: {
        id: 2,
        name: Relation Target 002,
        waft_id: 2,
      },
      relation_list: [],
      string: Delete me,
    },
    error: null,
  },
  requestId: 4,
}
`;

exports[`studio with default schema.prisma filename can respond to \`findMany\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: [
      {
        datetime: 2020-08-03T00:00:00.000Z,
        float: 3.14,
        id: 1,
        int: 42,
        relation: {
          id: 1,
          name: Relation Target 001,
          waft_id: 1,
        },
        relation_list: [],
        string: Some string,
      },
      {
        datetime: 1970-01-01T00:00:00.000Z,
        float: 0,
        id: 2,
        int: 0,
        relation: {
          id: 2,
          name: Relation Target 002,
          waft_id: 2,
        },
        relation_list: [],
        string: Delete me,
      },
    ],
    error: null,
  },
  requestId: 1,
}
`;

exports[`studio with default schema.prisma filename can respond to \`update\` queries 1`] = `
{
  action: clientRequest,
  channel: -prisma,
  payload: {
    data: {
      datetime: 2025-08-03T00:00:00.000Z,
      float: 100.5,
      id: 1,
      int: 100,
      relation: {
        id: 3,
        name: Relation Target 003,
        waft_id: 1,
      },
      relation_list: [
        {
          id: 3,
          name: Relation Target 003,
          waft_id: 1,
        },
      ],
      string: Changed String,
    },
    error: null,
  },
  requestId: 3,
}
`;
